File: k9mail/src/main/java/com/fsck/k9/message/AutocryptStatusInteractor.java
package com.fsck.k9.message;


import java.io.InputStream;

import android.app.PendingIntent;
import android.content.Intent;
import android.support.annotation.Nullable;
import android.support.annotation.VisibleForTesting;
import android.support.annotation.WorkerThread;

import org.openintents.openpgp.OpenPgpError;
import org.openintents.openpgp.util.OpenPgpApi;
import timber.log.Timber;


public class AutocryptStatusInteractor {
    private static final AutocryptStatusInteractor INSTANCE = new AutocryptStatusInteractor();

    public static AutocryptStatusInteractor getInstance() {
        return INSTANCE;
    }


    @WorkerThread
    public RecipientAutocryptStatus retrieveCryptoProviderRecipientStatus(
            OpenPgpApi openPgpApi, String[] recipientAddresses) {
        Intent intent = new Intent(OpenPgpApi.ACTION_QUERY_AUTOCRYPT_STATUS);
        intent.putExtra(OpenPgpApi.EXTRA_USER_IDS, recipientAddresses);

        Intent result = openPgpApi.executeApi(intent, (InputStream) null, null);

        switch (result.getIntExtra(OpenPgpApi.RESULT_CODE, OpenPgpApi.RESULT_CODE_ERROR)) {
            case OpenPgpApi.RESULT_CODE_SUCCESS:
                RecipientAutocryptStatusType type = getRecipientAutocryptStatusFromIntent(result);
                PendingIntent pendingIntent = result.getParcelableExtra(OpenPgpApi.RESULT_INTENT);
                return new RecipientAutocryptStatus(type, pendingIntent);

            case OpenPgpApi.RESULT_CODE_ERROR:
                OpenPgpError error = result.getParcelableExtra(OpenPgpApi.RESULT_ERROR);
                if (error != null) {
                    Timber.w("OpenPGP API Error #%s: %s", error.getErrorId(), error.getMessage());
                } else {
                    Timber.w("OpenPGP API Unknown Error");
                }
                return new RecipientAutocryptStatus(RecipientAutocryptStatusType.ERROR, null);
            case OpenPgpApi.RESULT_CODE_USER_INTERACTION_REQUIRED:
                // should never happen, so treat as error!
            default:
                return new RecipientAutocryptStatus(RecipientAutocryptStatusType.ERROR, null);
        }
    }

    @Nullable
    private RecipientAutocryptStatusType getRecipientAutocryptStatusFromIntent(Intent result) {
        boolean allKeysConfirmed = result.getBooleanExtra(OpenPgpApi.RESULT_KEYS_CONFIRMED, false);
        int autocryptStatus =
                result.getIntExtra(OpenPgpApi.RESULT_AUTOCRYPT_STATUS, OpenPgpApi.AUTOCRYPT_STATUS_UNAVAILABLE);

        switch (autocryptStatus) {
            case OpenPgpApi.AUTOCRYPT_STATUS_UNAVAILABLE:
                return RecipientAutocryptStatusType.UNAVAILABLE;
            case OpenPgpApi.AUTOCRYPT_STATUS_DISCOURAGE:
                if (allKeysConfirmed) {
                    return RecipientAutocryptStatusType.DISCOURAGE_CONFIRMED;
                } else {
                    return RecipientAutocryptStatusType.DISCOURAGE_UNCONFIRMED;
                }
            case OpenPgpApi.AUTOCRYPT_STATUS_AVAILABLE:
                if (allKeysConfirmed) {
                    return RecipientAutocryptStatusType.AVAILABLE_CONFIRMED;
                } else {
                    return RecipientAutocryptStatusType.AVAILABLE_UNCONFIRMED;
                }
            case OpenPgpApi.AUTOCRYPT_STATUS_MUTUAL:
                if (allKeysConfirmed) {
                    return RecipientAutocryptStatusType.RECOMMENDED_CONFIRMED;
                } else {
                    return RecipientAutocryptStatusType.RECOMMENDED_UNCONFIRMED;
                }
        }

        throw new IllegalStateException("encountered bad autocrypt status number!");
    }

    public static class RecipientAutocryptStatus {
        public final RecipientAutocryptStatusType type;
        public final PendingIntent intent;

        @VisibleForTesting
        public RecipientAutocryptStatus(RecipientAutocryptStatusType type, PendingIntent intent) {
            this.type = type;
            this.intent = intent;
        }

        public boolean hasPendingIntent() {
            return intent != null;
        }
    }

    public enum RecipientAutocryptStatusType {
        NO_RECIPIENTS (false, false, false),
        UNAVAILABLE (false, false, false),
        DISCOURAGE_UNCONFIRMED (true, false, false),
        DISCOURAGE_CONFIRMED (true, true, false),
        AVAILABLE_UNCONFIRMED (true, false, false),
        AVAILABLE_CONFIRMED (true, true, false),
        RECOMMENDED_UNCONFIRMED (true, false, true),
        RECOMMENDED_CONFIRMED (true, true, true),
        ERROR (false, false, false);

        private final boolean canEncrypt;
        private final boolean isConfirmed;
        private final boolean isMutual;

        RecipientAutocryptStatusType(boolean canEncrypt, boolean isConfirmed, boolean isMutual) {
            this.canEncrypt = canEncrypt;
            this.isConfirmed = isConfirmed;
            this.isMutual = isMutual;
        }

        public boolean canEncrypt() {
            return canEncrypt;
        }

        public boolean isConfirmed() {
            return isConfirmed;
        }

        public boolean isMutual() {
            return isMutual;
        }
    }
}

File: legacy/core/src/main/java/com/fsck/k9/message/AutocryptStatusInteractor.kt
package com.fsck.k9.message

import android.app.PendingIntent
import android.content.Intent
import androidx.annotation.VisibleForTesting
import androidx.annotation.WorkerThread
import androidx.core.content.IntentCompat
import java.io.InputStream
import net.thunderbird.core.logging.legacy.Log.w
import org.openintents.openpgp.OpenPgpError
import org.openintents.openpgp.util.OpenPgpApi

class AutocryptStatusInteractor {
    @WorkerThread
    fun retrieveCryptoProviderRecipientStatus(
        openPgpApi: OpenPgpApi,
        recipientAddresses: Array<String>,
    ): RecipientAutocryptStatus {
        val intent = Intent(OpenPgpApi.ACTION_QUERY_AUTOCRYPT_STATUS)
        intent.putExtra(OpenPgpApi.EXTRA_USER_IDS, recipientAddresses)

        val result = openPgpApi.executeApi(intent, null as InputStream?, null)

        return when (result.getIntExtra(OpenPgpApi.RESULT_CODE, OpenPgpApi.RESULT_CODE_ERROR)) {
            OpenPgpApi.RESULT_CODE_SUCCESS -> {
                val type = getRecipientAutocryptStatusFromIntent(result)
                val pendingIntent = IntentCompat.getParcelableExtra(
                    result,
                    OpenPgpApi.RESULT_INTENT,
                    PendingIntent::class.java,
                )
                RecipientAutocryptStatus(type, pendingIntent)
            }

            OpenPgpApi.RESULT_CODE_ERROR -> {
                val error = IntentCompat.getParcelableExtra(
                    result,
                    OpenPgpApi.RESULT_ERROR,
                    OpenPgpError::class.java,
                )
                if (error != null) {
                    w("OpenPGP API Error #%s: %s", error.getErrorId(), error.getMessage())
                } else {
                    w("OpenPGP API Unknown Error")
                }
                RecipientAutocryptStatus(RecipientAutocryptStatusType.ERROR, null)
            }

            OpenPgpApi.RESULT_CODE_USER_INTERACTION_REQUIRED -> RecipientAutocryptStatus(

                RecipientAutocryptStatusType.ERROR,
                null,
            )

            else -> RecipientAutocryptStatus(RecipientAutocryptStatusType.ERROR, null)
        }
    }

    private fun getRecipientAutocryptStatusFromIntent(result: Intent): RecipientAutocryptStatusType {
        val allKeysConfirmed = result.getBooleanExtra(OpenPgpApi.RESULT_KEYS_CONFIRMED, false)
        val autocryptStatus =
            result.getIntExtra(OpenPgpApi.RESULT_AUTOCRYPT_STATUS, OpenPgpApi.AUTOCRYPT_STATUS_UNAVAILABLE)

        return when (autocryptStatus) {
            OpenPgpApi.AUTOCRYPT_STATUS_UNAVAILABLE -> {
                RecipientAutocryptStatusType.UNAVAILABLE
            }

            OpenPgpApi.AUTOCRYPT_STATUS_DISCOURAGE -> {
                if (allKeysConfirmed) {
                    RecipientAutocryptStatusType.DISCOURAGE_CONFIRMED
                } else {
                    RecipientAutocryptStatusType.DISCOURAGE_UNCONFIRMED
                }
            }

            OpenPgpApi.AUTOCRYPT_STATUS_AVAILABLE -> {
                if (allKeysConfirmed) {
                    RecipientAutocryptStatusType.AVAILABLE_CONFIRMED
                } else {
                    RecipientAutocryptStatusType.AVAILABLE_UNCONFIRMED
                }
            }

            OpenPgpApi.AUTOCRYPT_STATUS_MUTUAL -> {
                return if (allKeysConfirmed) {
                    RecipientAutocryptStatusType.RECOMMENDED_CONFIRMED
                } else {
                    RecipientAutocryptStatusType.RECOMMENDED_UNCONFIRMED
                }
            }

            else -> error("encountered bad autocrypt status number!")
        }
    }

    data class RecipientAutocryptStatus @VisibleForTesting constructor(
        val type: RecipientAutocryptStatusType,
        val intent: PendingIntent?,
    ) {
        fun hasPendingIntent() = intent != null
    }

    enum class RecipientAutocryptStatusType(
        val canEncrypt: Boolean,
        val isConfirmed: Boolean,
        val isMutual: Boolean,
    ) {
        NO_RECIPIENTS(false, false, false),
        UNAVAILABLE(false, false, false),

        DISCOURAGE_UNCONFIRMED(true, false, false),
        DISCOURAGE_CONFIRMED(true, true, false),

        AVAILABLE_UNCONFIRMED(true, false, false),
        AVAILABLE_CONFIRMED(true, true, false),

        RECOMMENDED_UNCONFIRMED(true, false, true),
        RECOMMENDED_CONFIRMED(true, true, true),

        ERROR(false, false, false),
    }

    companion object {
        val instance: AutocryptStatusInteractor = AutocryptStatusInteractor()
    }
}
